### DISCUSSION TITLE
#(private field)에 관련된 질문입니다. 여러분은 어떤걸 선호하시고, 그 이유는 무엇일까요?

### DISCUSSION THREAD

**Discussion by lokba** [woowacourse-study/2022-code-review-study#14] [Lotto] [answered]
전제 : Controller에서 Model, View의 인스턴스를 생성해야하는 상황.

```javascript
//1번
class Controller{
  constructor(){
    this.model = new Model();
    this.view = new View();
  }
}
```

```javascript
//2번
class Controller{
  #model;
  #view;
 
  constructor(){
    this.#model = new Model();
    this.#view = new View();
  }
}

```

```javascript
//3번
class Controller{
  #model = null;
  #view = null;
 
  constructor(){
    this.#model = new Model();
    this.#view = new View();
  }
}

```

```javascript
//4번
class Controller{
  #model = new Model();
  #view = new View();
 
  constructor(){...}
}

```

**Comment by moonheekim0118** [comment]
저는 4번의 방식을 선호하는데요 이유는  멤버변수 #model과 #view 는 생성자 매개변수에 따라서 변할 일이 없기 때문입니다.
만약 생성자 매개변수에 의존한다면, 즉 아래처럼 생성되는 인스턴스에 따라서 #model 과 #view 멤버변수가 변한다면 생성자에서 선언해주는게 맞다고 보는데요, 그렇지 않으니까 저는 4번의 방식대로 하는 편입니다 ~ 다른분들은 어떨지 궁금하네요 !
그리고 , 3번의 경우는 굳이 null 로 초기화 해야할까요..? 어차피 undefined로 알아서 초기화 되는걸로  알고 있습니다.
```javascript
class Controller{
  #model;
  #view;
 
  constructor(ModelInstance, ViewInstance){
    this.#model = ModelInstance;
    this.#view = ViewInstance;
  }
}
```

**Comment by byhhh2** [comment]
저는 1번의 방식을 선호합니다.
사실상 `private`의 존재는 캡슐화를 위해서고, 캡슐화는 외부에 불필요한 정보를 노출시키지 않기 위해서 존재한다고 생각합니다. 
외부로 향하는 불필요한 정보는 side effect를 발생시키기 쉬우니깐요

그런데 controller가 가진 model과 view를 변경하려고 시도하는 협업자는 없다고 생각합니다..그렇기 때문에 저는 1번입니다. ㅎ

**Comment by wonsss** [comment]
저는 controller에서 model과 view의 인스턴스를 생성하는 상황에서는
1번의 방식을 사용함으로써 model과 view의 인스턴스에 `private`을 적용하지 않았습니다. 
그 이유는 일단 제가 만든 코드에서는 controller에 의존하고 있는 클래스가 없어서, 
controller가 가진 model과 view 인스턴스에 접근할 수 있는 주체가 없었기 때문입니다.
만약 controller를 의존하는 다른 클래스가 있고, 설계상 캡슐화를 해야 할 필요가 있다면, 4번의 방식을 활용할 것 같습니다.

그리고 그냥 클래스에서 메서드들을 캡슐화할 때는 4번의 방식을 활용하고 있습니다. 
prviate을 쓸 때 2번 또는 3번을 사용하는 이유가 무엇인지 궁금합니다. 기능상 서로 차이가 있나요?

**Comment by lokba** [comment, marked as answer]
모두의 의견은 상황에 따라 합리적이다 👍
